Server/src/main/java/org/gluu/oxauth/service/CibaRequestsProcessorJob.java

```diff
@@ -0,0 +1,169 @@
+import org.gluu.service.timer.schedule.TimerSchedule;
+import org.slf4j.Logger;
+
+import javax.ejb.DependsOn;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Event;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * @author Milton BO
+ * @version May 20, 2020
+ */
+@ApplicationScoped
+@DependsOn("appInitializer")
+@Named
+public class CibaRequestsProcessorJob {
+
+	@Inject
+	private Logger log;
+
+	@Inject
+	private AppConfiguration appConfiguration;
+
+	@Inject
+	private Event<TimerEvent> processorEvent;
+
+	@Inject
+	private CIBAPushErrorProxy cibaPushErrorProxy;
+
+	@Inject
+	private CIBAPingCallbackProxy cibaPingCallbackProxy;
+
+	@Inject
+	private AuthorizationGrantList authorizationGrantList;
+
+	@Inject
+	private GrantService grantService;
+
+	private long lastFinishedTime;
+
+	private AtomicBoolean isActive;
+
+	/**
+	 * Method invoked from the appInitializer to start processing every some time.
+	 */
+	public void initTimer() {
+		log.debug("Initializing CIBA requests processor");
+		this.isActive = new AtomicBoolean(false);
+		int intervalSec = appConfiguration.getBackchannelRequestsProcessorJobIntervalSec();
+
+		// Schedule to start processor every N seconds
+		processorEvent.fire(new TimerEvent(new TimerSchedule(intervalSec, intervalSec),
+				new CibaRequestsProcessorEvent(), Scheduled.Literal.INSTANCE));
+
+		this.lastFinishedTime = System.currentTimeMillis();
+	}
+
+	@Asynchronous
+	public void process(@Observes @Scheduled CibaRequestsProcessorEvent cibaRequestsProcessorEvent) {
+		if (this.isActive.get()) {
+			return;
+		}
+		if (!this.isActive.compareAndSet(false, true)) {
+			return;
+		}
+
+		try {
+			if (jobIsFree()) {
+				processImpl();
+				this.lastFinishedTime = System.currentTimeMillis();
+			} else {
+				log.trace("Starting conditions aren't reached for CIBA requestes processor");
+			}
+		} finally {
+			this.isActive.set(false);
+		}
+	}
+
+	/**
+	 * Defines whether the job is still in process or it is free according to the time interval defined.
+	 * @return True in case it is free to start a new process.
+	 */
+	private boolean jobIsFree() {
+		int interval = appConfiguration.getBackchannelRequestsProcessorJobIntervalSec();
+		if (interval < 0) {
+			log.info("CIBA Requests processor timer is disabled.");
+			log.warn("CIBA Requests processor timer Interval (cleanServiceInterval in oxauth configuration) is negative which turns OFF internal clean up by the server. Please set it to positive value if you wish internal CIBA Requests processor up timer run.");
+			return false;
+		}
+
+		long timeDiffrence = System.currentTimeMillis() - this.lastFinishedTime;
+		return timeDiffrence >= interval * 1000;
+	}
+
+	/**
+	 * Main process that process CIBA requests in cache.
+	 */
+	public void processImpl() {
+        try {
```

Comment:
This method means that whole oxauth can process only one single CIBA request per given period of time. Imagine we code Authorization Code Flow  in this way. We will get 1 request per second instead of hundreds. Yes, it's stable 1 req/s but we should handle many concurrently.

(Between, can you correct code formatter in your IDE to code that we already have? It would be easier to read and you will get less merging conflicts.)

Reply:
Actually, there is an instance in Cache which is a Map containing all `AuthReqIds` in `PENDING` state, so here I'm getting that map and process all values inside it, for instance, we could have ten requests in pending state, some of them could be answered, others could be still pending, so this method is responsible to update them according to the case. Once the request is updated, it's removed from this Map, therefore no more tasks are needed here for that request.

Reply:
I meant that it should be multi-threaded, not single thread job as it is in the PR (means ten requests will be handled by same single thread). Again, think about Authorization Code Flow.